pepr/privacy/art_inference_wrapper.py

```diff
@@ -0,0 +1,974 @@
+    pars_descriptors : dict
+        Dictionary of attack parameters and their description shown in the attack
+        report.
+        Example: {"attack_model_type": "Attack model type"} for the attribute named
+        "attack_model_type" of MembershipInferenceBlackBox.
+    attack_results : dict
+        Dictionary storing the attack model results.
+
+        * membership (list): List holding the inferred membership status, 1 indicates a
+          member and 0 indicates non-member per target model.
+        * tn (list): Number of true negatives per target model.
+        * tp (list): Number of true positives per target model.
+        * fn (list): Number of false negatives per target model.
+        * fp (list): Number of false positives per target model.
+        * precision (list): Attack precision per target model.
+        * recall (list): Attack recall per target model.
+        * accuracy (list): Attack accuracy per target model.
+    """
+
+    def __init__(
+        self,
+        attack_alias,
+        data,
+        labels,
+        data_conf,
+        target_models,
+        inference_attacks,
+        pars_descriptors,
+    ):
+        super().__init__(
+            attack_alias,
+            {},
+            data,
+            labels,
+            data_conf,
+            target_models,
+        )
+
+        self.inference_attacks = inference_attacks
+        self.pars_descriptors = pars_descriptors
+
+    @abstractmethod
+    def inference_run(self, tm_index):
+        """
+        Run an attack procedure.
+
+        Parameters
+        ----------
+        tm_index : int
+            Index of the currently attacked target model.
+
+        Returns
+        -------
+        An array holding the inferred membership status, 1 indicates a member and 0
+        indicates non-member, or class probabilities.
+        """
+        logger.error(
+            "The base class does not implement an attack! Please use an "
+            "attack class like 'MembershipInferenceBlackBox'."
+        )
+
+    def run(self):
+        """
+        Run ART inference attack
+        """
+        membership = []
+        acc = []
+        tn_list = []
+        tp_list = []
+        fn_list = []
+        fp_list = []
+        precision_list = []
+        recall_list = []
+
+        for i, inference_attack in enumerate(self.inference_attacks):
+            logger.info(f"Attack target model ({i + 1}/{len(self.inference_attacks)}).")
+
+            membership_results = self.inference_run(i)
+
+            true_mem = self.data_conf["attack_membership_status"][i]
+
+            # Evaluate attack model predictions
+            tn = np.count_nonzero((membership_results == 0) & (true_mem == False))
+            tp = np.count_nonzero((membership_results == 1) & (true_mem == True))
+            fn = np.count_nonzero((membership_results == 0) & (true_mem == True))
+            fp = np.count_nonzero((membership_results == 1) & (true_mem == False))
```

Comment:
Let's use ``is`` instead of ``==`` 😄

Reply:
It must be `==` or `np.equal` because it is an element-wise comparison with a numpy array. Python's `is` keyword cannot be used for that.

[`numpy.equal` doc](https://numpy.org/doc/stable/reference/generated/numpy.equal.html):
> The == operator can be used as a shorthand for np.equal on ndarrays.

Reply:
Oh yeah, is see... I was too fast. 👍